src/socket/tcp.rs

```diff
@@ -1170,6 +1147,51 @@ impl<'a> TcpSocket<'a> {
         }
 
         if let Some(ack_number) = repr.ack_number {
+            // TODO: When flow control is implemented,
+            // refractor the following block within that implementation
+
+            // Detect and react to duplicate ACKs by:
+            // 1. Check if duplicate ACK and change self.local_rx_dup_acks accordingly
+            // 2. If exactly 3 duplicate ACKs recived, set for fast retransmit
+            // 3. Update the last received ACK (self.local_rx_last_ack)
+            match self.local_rx_last_ack {
+                // Duplicate ACK if payload empty and ACK doesn't move send window ->
+                // Increment duplicate ACK count and set for retransmit if we just recived
+                // the third duplicate ACK
+                Some(ref mut last_rx_ack) if
+                    repr.payload.len() == 0 &&
+                    *last_rx_ack == ack_number => {
+                    // Increment duplicate ACK count
+                    match self.local_rx_dup_acks.checked_add(1) {
+                        Some(n) => {
+                            self.local_rx_dup_acks = n;
+                            net_debug!("{}:{}:{}: received duplicate ACK for seq {} (duplicate nr {})",
+                                    self.meta.handle, self.local_endpoint, self.remote_endpoint,
+                                    ack_number, self.local_rx_dup_acks);
+                        },
+                        None => {
+                            self.local_rx_dup_acks = u8::max_value();
+                            net_debug!("{}:{}:{}: received duplicate ACK for seq {} (duplicate nr overflowed)",
+                                    self.meta.handle, self.local_endpoint, self.remote_endpoint, ack_number);
+                        }
+                    };
+
+                    if self.local_rx_dup_acks == 3 {
+                        self.timer.set_for_fast_retransmit();
+                        net_debug!("{}:{}:{}: set socket for fast retransmit",
```

Comment:
Nit: you're setting *timer* for fast retransmit; a better message is something like "starting fast retransmit".

Reply:
I agree with this, however the reason for the awkward debug log is that this action doesn't technically send a retransmit, but it sets the socket up to send one. When the timeout expires for a "normal" retransmit it says something similar to this, hence I used the same phrase.